src/widget/wfindonwebmenu.cpp

```diff
@@ -124,54 +38,16 @@ void WFindOnWebMenu::openInBrowser(Service service,
     }
 }
 
-bool WFindOnWebMenu::hasEntriesForTrack(const Track& track) {
-    return !(track.getArtist().isEmpty() &&
-            track.getAlbum().isEmpty() &&
-            track.getTitle().isEmpty());
-}
-
-void WFindOnWebMenu::populateFromTrackProperties(
-        const Track& track,
-        const QString& serviceTitle,
-        Service service) {
-    const QString artist = track.getArtist();
-    const QString trackTitle = track.getTitle();
-    const QString album = track.getAlbum();
-    auto pServiceMenu = make_parented<QMenu>(this);
-    pServiceMenu->setTitle(serviceTitle);
-    addMenu(pServiceMenu);
-    addSeparator();
-    if (!artist.isEmpty()) {
-        addActions(service,
-                artist,
-                pServiceMenu,
-                WFindOnWebMenu::TrackSearchProperties::Artist);
-    }
-    if (!trackTitle.isEmpty()) {
-        if (!artist.isEmpty()) {
-            const QString artistWithTrackTitle = composeSearchQuery(trackTitle, artist);
-            addActions(service,
-                    artistWithTrackTitle,
-                    pServiceMenu,
-                    WFindOnWebMenu::TrackSearchProperties::ArtistAndTitle);
-        }
-        addActions(service,
-                trackTitle,
-                pServiceMenu,
-                WFindOnWebMenu::TrackSearchProperties::Title);
-    }
-    if (!album.isEmpty()) {
-        if (!artist.isEmpty()) {
-            const QString artistWithAlbum = composeSearchQuery(album, artist);
-            addActions(service,
-                    artistWithAlbum,
-                    pServiceMenu,
-                    WFindOnWebMenu::TrackSearchProperties::ArtistAndAlbum);
-        } else {
-            addActions(service,
-                    album,
-                    pServiceMenu,
-                    WFindOnWebMenu::TrackSearchProperties::Album);
-        }
+void WFindOnWebMenu::openInBrowser(const QString& query,
+        const QString& queryType,
+        const QString& serviceUrl) {
+    QUrlQuery urlQuery;
+    urlQuery.addQueryItem("q", query);
+    urlQuery.addQueryItem("type", queryType);
+    QUrl url(serviceUrl);
+    url.setQuery(urlQuery);
```

Comment:
I find this implementation "leaky". The query parameters are service specific. Its only a coincidence that they are shared between all three services. IMO it should `void WFindOnWebMenu::openInBrowser(const QUrl& url)` and each service implementation should compose their own QUrl. 
There is already evidence of this leakage in the implementation of the DiscogsService: This `openInBrowser` overload with the `queryType` argument exists only because Discogs URL-Schema is different from the rest.

Reply:
Oh, yeah. That's right. At first I thought that `void WFindOnWebMenu::openInBrowser` can be overloaded for the services, but at the end it will be not the best case. So it is better for to services compose their URL's.

Thanks for this feedback. I will change it to `void WFindOnWebMenu::openInBrowser(const QUrl& url)` and every service will be composing their own :+1:

Reply:
I have added one composer method in WFindOnWeb in order to prevent duplication, so all the services who has basic pagination can use this method and no need to define new method for similar services. 

I have added a function only to Discogs in the anon namespace for the detailed search. Any other services has different query types can have their own composer functions. 

Sorry for the delay, I was trying to figure out Cover Art fetcher meanwhile. I just had a time to look at it.

If it is not the best case, I can solve them. For the first requested change, I am waiting for @daschuer feedback, about the single free function.

Reply:
> I have added one composer method in WFindOnWeb in order to prevent duplication, so all the services who has basic pagination can use this method and no need to define new method for similar services.

Even though DRY is an important principle, sometimes duplicated code makes sense. In this case it does because the actual query is still dependent on each service. Deduplicating them between different services doesn't make sense. See https://schneide.blog/2019/12/27/code-duplication-is-not-always-evil/

However, what does make sense is to deduplicate some of the code in each `WFindOnWebMenu` childclass constructor. You are essentially copy pasting the same `!.isEmpty() {menu->addAction}` shenanigans everywhere. It would be trivial to extract the few variables and then do the rest in the helper function.

Reply:
That's fair enough. Every service should have their own composer.

Also I will try to get rid of all the `!.isEmpty() {menu->addAction}`'s. 

What I roughly think about it. There should be one main function in `wfindonwebmenu` which can create all the actions by taking few parameters, these can be the `service menu` `defaultUrl` `Enumaration of the Track Properties` (I'm back again with these) `defaulturl` as a string. So by calling this function in services, we can create all the actions and get rid of the duplication.

Reply:
I don't think shifting everything into the base-class is correct either. At that point we're back where we started. Start simple. By far the most common piece of code something like this:
```cpp
pDiscogsMenu->addAction(composeActionText(tr("Album"), album),
  this,
  [this, discogsUrlAlbum] {
      openInBrowser(discogsUrlAlbum);
  });
```
Can  you reduce this to a single function call that still makes sense?

Reply:
Yeap, I think I can reduce this simply by

`addActionForService(pDiscogsMenu, actionText, serviceUrl)`


I have reduced all the block with this but as you said that wouldn't be 100% correct. I could also check if the track property is empty or not. The method was like this;

`createAction(pDiscogsMenu, track, kSearchUrl, WFindOnWebMenu::TrackSearchProperties::ArtistAndAlbum);`

Reply:
IMO the first one is sufficient. The second one is "leaky" again because then the capabilities of the child-classes are limited by the parent class.

Reply:
Oh okay, thanks for this feedback. I will only make this addActions in the parent class. But this time, I also need to check if the track property is empty in the sub menu class, would it be correct?

Reply:
Yes.

Reply:
We can further iterate once you implemented that.

Reply:
Sorry for the delay, I was working on the Cover Art Fetcher again. 

Now I have pushed 2 commits. First I have added the method for creating the actions, second commit is the composer for each service. 

I gave a break with the cover art so I will be focusing on this PR after that.